### PR TITLE
Add filter by prefix in address and arpnd tables

### LINKED ISSUE #117: I'd like to be able to see all the address in a specific subnet
**Is your feature request related to a problem? Please describe.**
I know that 10.176.4.0/24 exists and is valid, I'd like to see all the addresses that are in that subnet

**Describe the solution you'd like**
Route show shows me that the prefix is valid, I'd like addresses to allow a prefix
```
jpiet> route show namespace=single-attach_bgp_numbered  prefix='172.16.4.0/24'
                      namespace   hostname           vrf         prefix                      nexthopIps              oifs protocol        source  metric  ipvers   action               timestamp
17   single-attach_bgp_numbered     edge01       default  172.16.4.0/24  [169.254.253.1, 169.254.254.1]  [eth2.2, eth1.2]      186    10.0.0.100      20       4  forward 2020-04-27 17:18:47.808
36   single-attach_bgp_numbered     exit01       default  172.16.4.0/24        [10.127.0.8, 10.127.1.8]      [swp1, swp2]      bgp                    20       4  forward 2020-04-27 17:18:47.808
54   single-attach_bgp_numbered     exit01  internet-vrf  172.16.4.0/24                [169.254.254.10]          [swp5.4]      bgp                    20       4  forward 2020-04-27 17:18:47.808
76   single-attach_bgp_numbered     exit02       default  172.16.4.0/24      [10.127.0.10, 10.127.1.10]      [swp1, swp2]      bgp                    20       4  forward 2020-04-27 17:18:47.808
94   single-attach_bgp_numbered     exit02  internet-vrf  172.16.4.0/24                [169.254.253.10]          [swp5.4]      bgp                    20       4  forward 2020-04-27 17:18:47.808
115  single-attach_bgp_numbered   internet       default  172.16.4.0/24                 [169.254.127.1]            [swp1]      bgp                    20       4  forward 2020-04-27 17:18:47.808
132  single-attach_bgp_numbered     leaf01       default  172.16.4.0/24        [10.127.0.0, 10.127.1.0]      [swp1, swp2]      bgp                    20       4  forward 2020-04-27 17:18:47.808
153  single-attach_bgp_numbered     leaf02       default  172.16.4.0/24        [10.127.0.2, 10.127.1.2]      [swp1, swp2]      bgp                    20       4  forward 2020-04-27 17:18:47.808
174  single-attach_bgp_numbered     leaf03       default  172.16.4.0/24        [10.127.0.4, 10.127.1.4]      [swp1, swp2]      bgp                    20       4  forward 2020-04-27 17:18:47.808
195  single-attach_bgp_numbered     leaf04       default  172.16.4.0/24                              []          [vlan10]   kernel    172.16.4.1      20       4  forward 2020-04-27 17:18:47.808
219  single-attach_bgp_numbered  server104       default  172.16.4.0/24                              []            [eth1]   kernel  172.16.4.104      20       4  forward 2020-04-27 17:18:47.808
239  single-attach_bgp_numbered    spine01       default  172.16.4.0/24                    [10.127.0.7]            [swp4]      bgp                    20       4  forward 2020-04-27 17:18:47.808
263  single-attach_bgp_numbered    spine02       default  172.16.4.0/24                    [10.127.1.7]            [swp4]      bgp                    20       4  forward 2020-04-27 17:18:47.808
jpiet> address show namespace=single-attach_bgp_numbered  address='172.16.4.0'
Empty DataFrame
Columns: [namespace, hostname, ifname, ipAddressList, macaddr, ip6AddressList, state, timestamp]
Index: []
jpiet>
```

**Additional context**

### PR DESCRIPTION
**Pull request by claudiolor**
This PR fixes #117 providing a way to filter by the address prefix the records of the `address` and the `arpnd` tables.
The `prefix` parameter has been added in order retrieve all the addresses belonging to the specified subnet:
```
claudio> address show namespace='ospf-single' prefix="10.255.2.0/24"
      namespace   hostname ifname    ipAddressList            macaddr                ip6AddressList state      vrf                        timestamp
0   ospf-single     edge01   eth0  10.255.2.109/24  52:54:00:7e:18:ce                            []    up  default 2021-03-21 19:46:42.916000+01:00
1   ospf-single     exit01   eth0   10.255.2.88/24  52:54:00:ad:87:c3  [fe80::5054:ff:fead:87c3/64]    up     mgmt 2021-03-21 19:46:44.640000+01:00
2   ospf-single     exit02   eth0   10.255.2.41/24  52:54:00:bf:13:32  [fe80::5054:ff:febf:1332/64]    up     mgmt 2021-03-21 19:46:44.640000+01:00
3   ospf-single   internet   eth0  10.255.2.239/24  52:54:00:e4:b2:74  [fe80::5054:ff:fee4:b274/64]    up     mgmt 2021-03-21 19:46:44.640000+01:00
4   ospf-single     leaf01   eth0   10.255.2.23/24  52:54:00:f0:31:4f  [fe80::5054:ff:fef0:314f/64]    up     mgmt 2021-03-21 19:46:44.640000+01:00
5   ospf-single     leaf02   eth0  10.255.2.203/24  52:54:00:c9:a9:2a  [fe80::5054:ff:fec9:a92a/64]    up     mgmt 2021-03-21 19:46:44.640000+01:00
6   ospf-single     leaf03   eth0  10.255.2.254/24  52:54:00:9b:fd:47  [fe80::5054:ff:fe9b:fd47/64]    up     mgmt 2021-03-21 19:46:44.640000+01:00
7   ospf-single     leaf04   eth0   10.255.2.10/24  52:54:00:b5:0b:49   [fe80::5054:ff:feb5:b49/64]    up     mgmt 2021-03-21 19:46:44.640000+01:00
8   ospf-single  server101   eth0   10.255.2.78/24  52:54:00:86:a5:9c                            []    up  default 2021-03-21 19:46:42.700000+01:00
9   ospf-single  server102   eth0   10.255.2.68/24  52:54:00:f3:36:56                            []    up  default 2021-03-21 19:46:42.674000+01:00
10  ospf-single  server103   eth0  10.255.2.182/24  52:54:00:8f:f9:5f                            []    up  default 2021-03-21 19:46:42.700000+01:00
11  ospf-single  server104   eth0  10.255.2.219/24  52:54:00:91:59:80                            []    up  default 2021-03-21 19:46:42.700000+01:00
12  ospf-single    spine01   eth0  10.255.2.217/24  52:54:00:a0:1a:f4  [fe80::5054:ff:fea0:1af4/64]    up     mgmt 2021-03-21 19:46:44.640000+01:00
13  ospf-single    spine02   eth0   10.255.2.60/24  52:54:00:ae:03:dc   [fe80::5054:ff:feae:3dc/64]    up     mgmt 2021-03-21 19:46:44.640000+01:00
``` 
The `address` parameter keeps its role, which is filtering by an exact IP address:
```
claudio> address show namespace='ospf-single' address="10.255.2.109"
     namespace hostname ifname    ipAddressList            macaddr ip6AddressList state      vrf                        timestamp
0  ospf-single   edge01   eth0  10.255.2.109/24  52:54:00:7e:18:ce             []    up  default 2021-03-21 19:46:42.916000+01:00
```

The following steps should be completed:

- [x] Add `prefix` parameter in `address` table - CLI
- [x] Add `prefix` parameter in `arpnd` table - CLI
- [x] Add `prefix` parameter in `address` table - REST
- [x] Add `prefix` parameter in `arpnd` table - REST
- [x] Provide a test suite

### CODE REVIEW

File: suzieq/engines/pandas/address.py [hunk 1]
```diff
@@ -96,22 +82,64 @@ def get(self, **kwargs) -> pd.DataFrame:
...
-            for a in v4addr:
-                query_str += f'{prefix} ipAddressList.str.startswith("{a}") '
-                prefix = 'or'
-        if v6addr:
-            for a in v6addr:
-                query_str += f'{prefix} ip6AddressList.str.startswith("{a}") '
-                prefix = 'or'
+        filter_prefix = ''
+
+        # Address and prefix filtering are mutual exclusive
```
Comment (LucaNicosia) [inline suzieq/engines/pandas/address.py:90]: I think it is better if you do a test like 
`if len(addr) > 0 and len(prefix) > 0:
  raise error "Cannot use both commands"`

**Review by LucaNicosia — APPROVED** [review]

**Review by claudious96 — APPROVED** [review]

File: suzieq/cli/sqcmds/ArpndCmd.py [hunk 2]
```diff
@@ -39,8 +39,12 @@ def __init__(
               description="IP address, in quotes, to qualify output")
     @argument("macaddr",
               description="MAC address, in quotes, to qualify output")
+    @argument("prefix",
```
Comment (ddutt) [inline suzieq/cli/sqcmds/ArpndCmd.py:42]: The text string maybe too long to display well. Have you checked? Alternately, a shorter description might be "Show all addresses in this subnet prefix"

File: suzieq/cli/sqcmds/AddressCmd.py [hunk 3]
```diff
@@ -36,11 +36,18 @@ def __init__(
         )
 
     @command("show")
-    @argument("address", description="Address, in quotes, to show info for")
-    @argument("vrf", description="VRF to qualify the address")
-    @argument("ipvers", description="type of address, v4, v6 or l2",
+    @argument("address",
+              description="Address, in quotes, to show info for")
+    @argument("prefix",
```
Comment (ddutt) [inline suzieq/cli/sqcmds/AddressCmd.py:41]: See comment below about how long the help text is and if its an issue.

File: suzieq/engines/pandas/address.py [hunk 4]
```diff
@@ -37,14 +39,13 @@ def get(self, **kwargs) -> pd.DataFrame:
         addnl_fields = ['master']
         drop_cols = []
 
-        v4addr = []
-        v6addr = []
-        macaddr = []
-        try:
+        if len(prefix) > 0:
```
Comment (ddutt) [inline suzieq/engines/pandas/address.py:42]: You can do "if prefix", no need to check for length each time, its more pythonic

File: suzieq/engines/pandas/address.py [hunk 5]
```diff
@@ -96,22 +82,64 @@ def get(self, **kwargs) -> pd.DataFrame:
...
-                query_str += f'{prefix} ipAddressList.str.startswith("{a}") '
-                prefix = 'or'
-        if v6addr:
-            for a in v6addr:
-                query_str += f'{prefix} ip6AddressList.str.startswith("{a}") '
-                prefix = 'or'
+        filter_prefix = ''
+
+        # Address and prefix filtering are mutual exclusive
+        if len(addr) > 0:
```
Comment (ddutt) [inline suzieq/engines/pandas/address.py:91]: Use "if address" instead of len

File: suzieq/engines/pandas/address.py [hunk 6]
```diff
@@ -96,22 +82,64 @@ def get(self, **kwargs) -> pd.DataFrame:
...
+                    query_str += (f'{filter_prefix} '
+                                  f'ipAddressList.str.startswith("{a}") ')
+                    filter_prefix = 'or'
+            if v6addr:
+                for a in v6addr:
+                    query_str += (f'{filter_prefix} '
+                                  f'ip6AddressList.str.startswith("{a}") ')
+                    filter_prefix = 'or'
+
+        elif len(prefix) > 0:
```
Comment (ddutt) [inline suzieq/engines/pandas/address.py:126]: Change to "elif prefix"

File: suzieq/engines/pandas/engineobj.py [hunk 7]
```diff
@@ -75,6 +77,21 @@ def _handle_user_query_str(self, df: pd.DataFrame,
...
+    def _is_in_subnet(self, addr: PandasObject, net: str) -> PandasObject:
+        """Check if the IP addresses in a Pandas dataframe
+        belongs to the given subnet
+
+        Args:
+            addr (PandasObject): the collection of ip addresses to check
+            net: (str): network id of the subnet
+
+        Returns:
+            PandasObject: A collection of bool reporting the result
```
Comment (ddutt) [inline suzieq/engines/pandas/engineobj.py:89]: Isn't this returning a pd.Series?

File: suzieq/engines/pandas/engineobj.py [hunk 8]
```diff
@@ -75,6 +77,21 @@ def _handle_user_query_str(self, df: pd.DataFrame,
...
+        """
+        return addr.apply(lambda a: (
+            False if not a else ip_address(a.split("/")[0]) in ip_network(net))
```
Comment (ddutt) [inline suzieq/engines/pandas/engineobj.py:92]: To improve the efficiency of not computing the network each time, can you pull out the ```ip_network(net)``` to be outside this function and pass it the output of ```ip_network(net)``` ?

**Review by ddutt — CHANGES_REQUESTED** [review]

**Pull request merged**